### ISSUE TITLE
auto-formatting bucket paths

### ISSUE THREAD

**Issue by nick-youngblut** [nextflow-io/nextflow#4486]
## New feature

If a user provides a bucket path as a pipeline input parameter (e.g., `output_dir = gs://path/to/my/bucket`), then the path could end with a forward flash or not. If there is a forward slash, and code in the pipeline appends to the path as in `output_dir + /sub_dir/file.txt`, then the resulting bucket path is `gs://path/to/my/bucket//sub_dir/file.txt`. Nextflow would fail to publish this output file to GCP Cloud Storage, since the path contains "//" in the path. 

It would be helpful if Nextflow automatically checked and reformatted all bucket paths (e.g., as part of the publishDir function) to make sure that they are formatted correctly (e.g., `gs://path/to/my/bucket/sub_dir/file.txt` instead of `gs://path/to/my/bucket//sub_dir/file.txt`).

This should be a simple regex, with the stipulation that the "//" in the bucket prefix is maintained (e.g., `gs://`).

While such a regex solution can be implemented in each pipeline, this requires many developers to include the same boilerplate code in each pipeline.

## Usage scenario 

Whenever cloud storage bucket paths are provided by the user, and thus could contain `//+` in the path.

## Suggest implementation 

A simple regex that replaces all `/+` with `/`, while accounting for the bucket prefix (e.g., `gs://`).

**Comment by bentsherman** [comment]
I believe the best practice here is to do `file(output_dir).resolve('sub_dir/file.txt')` which will handle duplicate slashes

**Comment by nick-youngblut** [comment]
> I believe the best practice here is to do file(output_dir).resolve('sub_dir/file.txt') which will handle duplicate slashes

@bentsherman do you mean that pipeline developers should implement that code into all of their applicable pipelines, or that `file().resolve()` should be added to the Nextflow codebase to automatically handle duplicate slashes so that pipeline developers do not have to worry about potential issues with bucket paths?

**Comment by bentsherman** [comment]
Pipeline developers should do this, it is a best practice when working with file paths in pipeline code. In fact I believe it is documented [here](https://nextflow.io/docs/latest/script.html#opening-files)

**Comment by nick-youngblut** [comment]
> Pipeline developers should do this

This results in boilerplate code in many pipelines (with some code implementations better than others). Why not just implement it once in Nextflow and be done with it? There is no value in having multiple forward slashes in a path, so automatically correcting this issue would not cause problems for existing pipelines.

**Comment by bentsherman** [comment]
I don't consider it boilerplate code, it's a standard library feature. You can go even simpler:
```groovy
file(output_dir) / 'sub_dir/file.txt'
```

**Comment by bentsherman** [comment]
The docs seem pretty clear to me, since there is a note in that section about the correct way to resolve paths. But if you think it can be made clearer, feel free to submit a PR 😄

**Comment by nick-youngblut** [comment]
@bentsherman as an FYI: I've run into this issue with 2 pipelines just this week: 

* https://github.com/nf-core/differentialabundance 
* https://github.com/ScaleBio/ScaleRna

Both are not handling user-input file paths correctly, which results in double forward slashes in the output file bucket paths, and so there are `Failed to publish file` warnings.

**Comment by nick-youngblut** [comment]
@bentsherman I run into a situation where auto-formatting of bucket paths for `publishDir` would be REALLY helpful:

I want to set the base directory as a bucket if using the `gcp` profile, but a local directory if not using the `vm` profile:

```
vm {
   params.output_base  = "pipeline_results/"
}
gcp {
    params.output_base  = "gs://my_bucket/"
}
```

The user provides the rest of the path via `output_dir`. So, the entire path is created via `output_base` + `output_dir`. Example: "gs://my_bucket/" + "path/to/my_results".

However, there appears no straight forward way to prevent users from breaking the path via including an extra forward slash. For example: "gs://my_bucket/" + "/path/to/my_results" = "gs://my_bucket//path/to/my_results". 

`publishDir file(params.output_base) /  file(params.output_dir)` does not work. 

This also does not work, since it will strip off the second forward slash in the bucket prefix (e.g., "gs://" => "gs:/"):

```
import java.nio.file.Paths

publishDir Paths.get(params.output_base, params.output_dir)
```

Is there any "good" way of dynamically dealing with bucket paths in Nextflow, as described in my scenario above?

**Comment by bentsherman** [comment]
I think the correct syntax would be:
```groovy
publishDir file(params.output_base) / params.output_dir / "my_process_output_dir"
```

But also, I think we fixed the issue with GCS paths in [this commit](https://github.com/nextflow-io/nextflow/commit/18ec484f772c32a4a59982e1b7accd050c08f02b#diff-be4c220c7fcd4ba52b0dc3718a8d99f2b67d504731e2e4c25c475cfbad2f022a), which is included in 24.10

**Comment by nick-youngblut** [comment]
This still results in code coloring issues, at least for the VS code extension:

<img width="916" alt="Image" src="https://github.com/user-attachments/assets/7c15dc08-8901-4fd7-913e-cefbea104230" />
